### PR TITLE
Option "Show media posts with missing alt text"

### PR DESCRIPTION
**Pull request by Botiplz**
Based on: https://corteximplant.com/@Botiplz/110754125361384888

New option "Show media posts with missing alt text"
Hides Posts without alt text in "HOME, PUBLIC, ACCOUNT" and shows a warning in "THREAD, NOTIFICATIONS"

### CODE REVIEW

File: mastodon/src/main/java/org/joinmastodon/android/fragments/SettingsFragment.java
Comment (FineFindus) [inline mastodon/src/main/java/org/joinmastodon/android/fragments/SettingsFragment.java:1]: This is for the old settings UI, which is no longer used.

File: mastodon/src/main/res/values/strings_sk.xml
Comment (FineFindus) [inline mastodon/src/main/res/values/strings_sk.xml:1]: Moshidon specific strings should be in `strings_mo.xml`.  `strings_sk.xml` is for Megalodon-specific strings.

File: mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java [hunk 1]
```diff
@@ -72,9 +85,21 @@ public boolean test(Status status){
...
 				.findAny();
 
+		//Apply client filters if no server filter is triggered
+		if (applyingFilter.isEmpty()) {
+			applyingFilter = clientFilters.stream()
+					// only apply filters for given context
+					.filter(filter -> filter.context.contains(context))
+					// treating filterAction = null (from filters list) as FilterAction.HIDE
+					.filter(filter -> filter.filterAction == null ? action == HIDE : filter.filterAction == action)
+					//client filter has to match the status
```
Comment (FineFindus) [inline mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java:98]: Please use comments to add additional context to code. It is fairly obvious from the that the filter has to match the status.

File: mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java [hunk 2]
```diff
@@ -11,43 +20,47 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-public class StatusFilterPredicate implements Predicate<Status>{
+public class StatusFilterPredicate implements Predicate<Status> {
+
+	//Hide in timelines and warn in threads
+	private final List<Filter> clientFilters = List.of(	new AltTextFilter(HIDE, HOME, PUBLIC, ACCOUNT),
+														new AltTextFilter(WARN, THREAD, NOTIFICATIONS));
```
Comment (FineFindus) [inline mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java:27]: Please format this properly.

File: mastodon/src/main/java/org/joinmastodon/android/ui/displayitems/WarningFilteredStatusDisplayItem.java [hunk 3]
```diff
@@ -55,7 +51,11 @@ public Holder(Context context, ViewGroup parent){
 		@Override
 		public void onBind(WarningFilteredStatusDisplayItem item) {
 			filteredItems = item.filteredItems;
-			text.setText(item.parentFragment.getString(R.string.sk_filtered, item.applyingFilter.title));
+			if(item.applyingFilter instanceof AltTextFilter){
+				text.setText(item.parentFragment.getString(R.string.sk_filtered,item.parentFragment.getString(R.string.sk_no_alt_text)));
+			}else{
+				text.setText(item.parentFragment.getString(R.string.sk_filtered, item.applyingFilter.title));
+			}
```
Comment (FineFindus) [inline mastodon/src/main/java/org/joinmastodon/android/ui/displayitems/WarningFilteredStatusDisplayItem.java:58]:
```suggestion
String title = item.applyingFilter instanceof AltTextFilter ? item.parentFragment.getString(R.string.sk_no_alt_text) : applyingFilter.title;
text.setText(item.parentFragment.getString(R.string.sk_filtered, text));
```
Both branches share a larger amount of code, so it is easier to read when using a conditional.

File: mastodon/src/main/java/org/joinmastodon/android/fragments/SettingsFragment.java [hunk 4]
```diff
@@ -296,6 +296,10 @@ public void onCreate(Bundle savedInstanceState){
 			GlobalUserPreferences.showNoAltIndicator=i.checked;
 			GlobalUserPreferences.save();
 		}));
+		items.add(new SwitchItem(R.string.sk_settings_show_posts_without_alt, R.drawable.ic_fluent_eye_tracking_on_24_regular, GlobalUserPreferences.showPostsWithoutAlt, i->{
```
Comment (FineFindus) [inline mastodon/src/main/java/org/joinmastodon/android/fragments/SettingsFragment.java:299]: I think this setting should include a summary, explaining to the user what it does. What do you think @LucasGGamerM?

File: mastodon/src/main/java/org/joinmastodon/android/model/AltTextFilter.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+	public AltTextFilter(FilterAction filterAction, FilterContext firstContext, FilterContext... restContexts) {
+		this.filterAction = filterAction;
+		isRemote = false;
+		context = EnumSet.of(firstContext, restContexts);
+	}
+
+	@Override
+	public boolean matches(Status status) {
+		return !GlobalUserPreferences.showPostsWithoutAlt && status.getContentStatus().mediaAttachments.stream().map(attachment -> attachment.description).anyMatch(StringUtil::isBlank);
```
Comment (FineFindus) [inline mastodon/src/main/java/org/joinmastodon/android/model/AltTextFilter.java:19]: I think the preference should be done when applying the filter.

**Review by FineFindus — CHANGES_REQUESTED** [review]

File: mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java (hunk 2 above)
Comment (Botiplz) [reply mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java:27]: I didn't find a .editorconfig and saw different formattings between different files. Is there a project wide definition?

Comment (Botiplz) [reply mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java:27]: I didn't find a .editorconfig and saw different formattings between different files. Is there a project wide definition?

File: mastodon/src/main/java/org/joinmastodon/android/fragments/SettingsFragment.java
Comment (Botiplz) [reply mastodon/src/main/java/org/joinmastodon/android/fragments/SettingsFragment.java:1]: Reverted changes on SettingsFragment

File: mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java (hunk 1 above)
Comment (Botiplz) [reply mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java]: I copied the comments from above -> removed them

File: mastodon/src/main/java/org/joinmastodon/android/fragments/SettingsFragment.java (hunk 4 above)
Comment (Botiplz) [reply mastodon/src/main/java/org/joinmastodon/android/fragments/SettingsFragment.java]: summary added

File: mastodon/src/main/java/org/joinmastodon/android/model/AltTextFilter.java (hunk 5 above)
Comment (Botiplz) [reply mastodon/src/main/java/org/joinmastodon/android/model/AltTextFilter.java]: Updated the StatusFilterPredicate which sets the clientFilters based on the UserPreferences

File: mastodon/src/main/res/values/strings_sk.xml
Comment (Botiplz) [reply mastodon/src/main/res/values/strings_sk.xml:1]: already wondered about that - changed

File: mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java (hunk 2 above)
Comment (FineFindus) [reply mastodon/src/main/java/org/joinmastodon/android/utils/StatusFilterPredicate.java]: Not really. The official app often omits spaces when assigning variables, whereas we Megalodon, and we often don't. So as long as it looks fine, it's ok. The issue here was the tab spacing and the basically empty new line.

File: mastodon/src/main/res/values-de-rDE/strings_mo.xml [hunk 6]
```diff
@@ -75,4 +75,6 @@
     <string name="mo_setting_uniform_summary">App-Symbol für alle Benachrichtigungen verwenden</string>
     <string name="mo_setting_reduced_motion_summary">Animationen für Interaktionen deaktivieren</string>
     <string name="mo_setting_play_gif_summary">GIFs in Profilbildern und Emojis automatisch abspielen</string>
+    <string name="mo_settings_show_posts_without_alt">Zeige Medienbeiträge ohne Bildbeschreibung</string>
+    <string name="mo_settings_show_posts_without_alt_summary">Beiräge werden in allen Timelines ausgeblendet, können aber in Threads und Benachrichtungen ausgeklappt werden</string>
```
Comment (FineFindus) [inline mastodon/src/main/res/values-de-rDE/strings_mo.xml:79]:
```suggestion
    <string name="mo_settings_show_posts_without_alt_summary">Beiträge werden in allen Timelines ausgeblendet, können aber in Threads und Benachrichtungen angezeigt werden</string>
```
Typo, and I'm not sure about `ausgeklappt`. Maybe use `angezeigt` or `aufgedeckt` instead?

File: mastodon/src/main/res/values/strings_mo.xml [hunk 7]
```diff
@@ -87,7 +87,9 @@
     <string name="mo_setting_disable_swipe_summary">Swipe to change viewed timeline</string>
     <string name="mo_setting_haptic_feedback_summary">Vibrate when interacting with posts</string>
     <string name="mo_swap_bookmark_with_reblog_summary">Bookmark or reblog posts from the notification</string>
+    <string name="mo_settings_show_posts_without_alt_summary">Posts will be hidden in all timelines, but can be revealed in threads and notifications</string>
```
Comment (FineFindus) [inline mastodon/src/main/res/values/strings_mo.xml:90]: Not sure about the `all`, I think just timelines would be fine by itself, but I'm fine with it being either way.

**Review by FineFindus — APPROVED** [review]
Found a typo and added a few slight suggestions for wording, but I'm also fine with them staying like this. LGTM, otherwise, thanks for the pr.

File: mastodon/src/main/res/values/strings_mo.xml (hunk 7 above)
Comment (Botiplz) [reply mastodon/src/main/res/values/strings_mo.xml:90]: All/Alle should be updated in the German translation aswell, if someone else insists on it.

**Comment by Botiplz** [comment]
Misclicked, sorry (my first GitHub PR 👀)

**Comment by LucasGGamerM** [comment]
Some small health issues have stopped me from doing some maintenance things. I will review this soon. Thank you for the PR!

**Comment by Botiplz** [comment]
> Some small health issues have stopped me from doing some maintenance things. I will review this soon. Thank you for the PR!

No worries! Health comes first!

**Comment by LucasGGamerM** [comment]
This should be merged soon

**Comment by LucasGGamerM** [comment]
Such a pain. It compiles, but just doesnt want to work

**Comment by LucasGGamerM** [comment]
Its all fixed :D

**Pull request merged**